File: engine/ipm/src/main/java/com/motaword/ipm/engine/IPMProcessConfigCreator.java
package com.motaword.ipm.engine;

import com.motaword.ipm.engine.translator.controller.TooFrequentTranslation;
import com.typesafe.config.Config;
import pl.touk.nussknacker.engine.api.CustomStreamTransformer;
import pl.touk.nussknacker.engine.api.ProcessListener;
import pl.touk.nussknacker.engine.api.Service;
import pl.touk.nussknacker.engine.api.exception.ExceptionHandlerFactory;
import pl.touk.nussknacker.engine.api.process.SingleNodeConfig$;
import pl.touk.nussknacker.engine.api.process.SinkFactory;
import pl.touk.nussknacker.engine.api.process.SourceFactory;
import pl.touk.nussknacker.engine.api.process.WithCategories;
import pl.touk.nussknacker.engine.api.signal.ProcessSignalSender;
import pl.touk.nussknacker.engine.example.LoggingExceptionHandlerFactory;
import pl.touk.nussknacker.engine.javaapi.process.ExpressionConfig;
import pl.touk.nussknacker.engine.javaapi.process.ProcessConfigCreator;
import scala.collection.JavaConversions;

import java.util.*;

public class IPMProcessConfigCreator implements ProcessConfigCreator {

    @Override
    public Map<String, WithCategories<Service>> services(Config config) {
        return Collections.emptyMap();
    }

    @Override
    public Map<String, WithCategories<SourceFactory<?>>> sourceFactories(Config config) {
        return Collections.emptyMap();
    }

    @Override
    public Map<String, WithCategories<SinkFactory>> sinkFactories(Config config) {
        return Collections.emptyMap();
    }

    @Override
    public Map<String, WithCategories<CustomStreamTransformer>> customStreamTransformers(Config config) {
        Map<String, WithCategories<CustomStreamTransformer>> m = new HashMap<>();
        m.put("Translator.TooFrequentTranslation", all(new TooFrequentTranslation(), "Translator"));
        return m;
    }

    @Override
    public Map<String, WithCategories<ProcessSignalSender>> signals(Config config) {
        return Collections.emptyMap();
    }

    @Override
    public Collection<ProcessListener> listeners(Config config) {
        return Collections.emptyList();
    }

    @Override
    public ExceptionHandlerFactory exceptionHandlerFactory(Config config) {
        return new LoggingExceptionHandlerFactory(config);
    }

    @Override
    public ExpressionConfig expressionConfig(Config config) {
        return new ExpressionConfig(Collections.emptyMap(), Collections.emptyList());
    }

    @Override
    public Map<String, String> buildInfo() {
        return Collections.emptyMap();
    }

    private <T> WithCategories<T> all(T value, ArrayList<String> categories) {
        if (categories == null) {
            categories = new ArrayList<>();
        }

        return new WithCategories<>(
                value,
                JavaConversions.collectionAsScalaIterable(categories).toList(),
                SingleNodeConfig$.MODULE$.zero());
    }

    private <T> WithCategories<T> all(T value, String category) {
        if (category == null) {
            category = "Generic";
        }

        ArrayList<String> categories = new ArrayList<>();
        categories.add(category);

        return all(value, categories);
    }

}

File: engine/ipm/src/main/scala/com/motaword/ipm/engine/IPMProcessConfigCreator.scala
package com.motaword.ipm.engine

import com.motaword.ipm.engine.translator.controller.TooFrequentTranslation
import com.typesafe.config.Config
import pl.touk.nussknacker.engine.api._
import pl.touk.nussknacker.engine.api.exception.{EspExceptionHandler, ExceptionHandlerFactory}
import pl.touk.nussknacker.engine.api.process._
import pl.touk.nussknacker.engine.api.signal.ProcessSignalSender

class IPMProcessConfigCreator extends ProcessConfigCreator {

  override def customStreamTransformers(config: Config): Map[String, WithCategories[CustomStreamTransformer]] = {
    Map(
      "TooFrequentTranslation" -> WithCategories(new TooFrequentTranslation, "Translator")
    )
  }

  override def services(config: Config): Map[String, WithCategories[Service]] =
    Map.empty

  override def sourceFactories(config: Config): Map[String, WithCategories[SourceFactory[_]]] =
    Map.empty

  override def sinkFactories(config: Config): Map[String, WithCategories[SinkFactory]] =
    Map.empty

  override def listeners(config: Config): Seq[ProcessListener] =
    Nil

  //TODO: this does not work for Flink procsses -> as it is doesn't define restart strategy...
  override def exceptionHandlerFactory(config: Config): ExceptionHandlerFactory =
    ExceptionHandlerFactory.noParams(_ => EspExceptionHandler.empty)

  override def expressionConfig(config: Config) = ExpressionConfig(Map.empty, List.empty)

  override def buildInfo(): Map[String, String] = {
    val engineBuildInfo = pl.touk.nussknacker.engine.version.BuildInfo.toMap.map { case (k, v) => s"engine-$k" -> v.toString }
    engineBuildInfo ++ Map(
      "process-version" -> "0.1"
    )
  }

  override def signals(config: Config): Map[String, WithCategories[ProcessSignalSender]] =
    Map.empty


}
